torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+import torch
+from collections import defaultdict
+from torch._six import container_abcs
+
+
+class _MultiDeviceReplicator(object):
+    """
+    Lazily serves copies of a tensor to requested devices.  Copies are cached per-device.
+    """
+    def __init__(self, master_tensor):
+        assert master_tensor.is_cuda
+        self.master = master_tensor
+        self._per_device_tensors = {}
+
+    def get(self, device):
+        retval = self._per_device_tensors.get(device, None)
+        if retval is None:
+            retval = self.master.to(device=device, non_blocking=True, copy=True)
+            self._per_device_tensors[device] = retval
+        return retval
+
+
+class AmpScaler(object):
+    """
+    :class:`AmpScaler` performs dynamic gradient scaling.
+
+    Here's how that looks in a simple example::
+
+        # Create an AmpScaler instance.
+        scaler = AmpScaler()
+        ...
+        for input, target in data:
+            optimizer.zero_grad()
+            output = model(input)
+            loss = loss_fn(output, target)
+
+            # Scale the loss, and call backward() on the scaled loss to create scaled gradients.
+            scaler.scale(loss).backward()
+
+            # Carry out a scaling-safe step.  scaler.step() unscales the optimizer's gradients
+            # and skips optimizer.step() if the gradients contain infs or NaNs.
+            scaler.step(optimizer)
+
+            # Update the scale for next iteration.
+            scaler.update()
+
+    See the :ref:`Gradient Scaling Examples<gradient-scaling-examples>` for usage in more complex cases.
+
+    ``scaler`` maintains the scale factor internally.  To leverage ``float16``'s full dynamic range,
+    ``scaler`` attempts to use the largest scale factor it can without incurring overflow.
+    It does so by checking the gradients for infs and NaNs during every :meth:`step` or separate :meth:`unscale`.
+    If no infs/NaNs are found, :meth:`step` runs the underlying ``optimizer.step()`` as usual and
+    :meth:`update` multiplies the scale factor by the growth factor.  If infs/NaNs are found, :meth:`step` skips the
+    underlying ``optimizer.step()`` (so the params themselves remain uncorrupted) and multiplies the scale factor by
+    the backoff factor.
+
+    The scale factor often causes infs/NaNs to appear in gradients for the first few iterations as its
```

Comment:
Pls kill this paragraph.

Greg: <nodding>

Reply:
This is one of the most common things people ask about.  They start training, see inf/nans immediately, and file an issue.  I'd rather have a few lines of documentation to anticipate their question.